### ISSUE TITLE
Mismatched shapes when ComputationGraph#output is called within iterationDone() of IterationListener

### ISSUE THREAD

**Issue by fahman** [deeplearning4j/deeplearning4j#3055]
In certain conditions calling CompuationGraph#output(false, data) causes following exception:

`java.lang.IllegalStateException: Mismatched shapes (shape = [2, 2], row vector shape =[4, 1])
	at org.nd4j.linalg.api.ndarray.BaseNDArray.doColumnWise(scratch_13.scala:2171)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.muliColumnVector(scratch_13.scala:2583)
	at org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.activateHelper(scratch_13.scala:245)
	at org.deeplearning4j.nn.layers.recurrent.GravesBidirectionalLSTM.activateOutput(scratch_13.scala:183)
	at org.deeplearning4j.nn.layers.recurrent.GravesBidirectionalLSTM.activate(scratch_13.scala:170)
	at org.deeplearning4j.nn.graph.vertex.impl.LayerVertex.doForward(scratch_13.scala:99)
	at org.deeplearning4j.nn.graph.ComputationGraph.feedForward(scratch_13.scala:1112)
	at org.deeplearning4j.nn.graph.ComputationGraph.feedForward(scratch_13.scala:1082)
	at org.deeplearning4j.nn.graph.ComputationGraph.output(scratch_13.scala:1167)
	at #worksheet#.OutputListener.iterationDone(scratch_13.scala:60)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(scratch_13.scala:65)
	at org.deeplearning4j.optimize.Solver.optimize(scratch_13.scala:47)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(scratch_13.scala:868)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(scratch_13.scala:684)
	at #worksheet#.#worksheet#(scratch_13.scala:95)
`

Conditions are:
- Config with GravesBidirectionalLSTM is used;
- Masking is used;
- Training is done in minibatches of size X;
- ComputationGraph#output(false, testSample) is called where testSample has different size than X (e.g. 2 samples instead of 10 samples during training);
- output() is called within IterationListener#iterationDone(Model, Int)

The fully functioning snippet bellow (sorry for my scala):

```
import java.util

import org.deeplearning4j.nn.api.{Model, OptimizationAlgorithm}
import org.deeplearning4j.nn.conf.graph.rnn.{DuplicateToTimeSeriesVertex, LastTimeStepVertex}
import org.deeplearning4j.nn.conf.layers.{DenseLayer, GravesBidirectionalLSTM, RnnOutputLayer}
import org.deeplearning4j.nn.conf.{NeuralNetConfiguration, Updater}
import org.deeplearning4j.nn.graph.ComputationGraph
import org.deeplearning4j.nn.weights.WeightInit
import org.deeplearning4j.optimize.api.IterationListener
import org.nd4j.linalg.activations.Activation
import org.nd4j.linalg.api.ndarray.INDArray
import org.nd4j.linalg.dataset.DataSet
import org.nd4j.linalg.dataset.api.DataSetPreProcessor
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator
import org.nd4j.linalg.factory.Nd4j
import org.nd4j.linalg.indexing.NDArrayIndex
import org.nd4j.linalg.lossfunctions.LossFunctions.LossFunction


val ndArray = Nd4j.create(Array(1.0, 2.0, 3.0, 4.0, 5.0, 6.0)).reshape(6, 1)
val ndarr = Nd4j.create(4, 1, 6)
for (i <- 0 to 5) ndarr.put(Array(NDArrayIndex.point(0), NDArrayIndex.all(), NDArrayIndex.point(i)), ndArray.getRow(i))
for (i <- 0 to 5) ndarr.put(Array(NDArrayIndex.point(1), NDArrayIndex.all(), NDArrayIndex.point(i)), ndArray.getRow(5 - i))
for (i <- 0 to 5) ndarr.put(Array(NDArrayIndex.point(2), NDArrayIndex.all(), NDArrayIndex.point(i)), ndArray.getRow(i))
for (i <- 0 to 5) ndarr.put(Array(NDArrayIndex.point(3), NDArrayIndex.all(), NDArrayIndex.point(i)), ndArray.getRow(5 - i))

class BatchedIterator(val batch: Int) extends DataSetIterator {
  override def next(): DataSet = next(batch)
  override def next(batchSize: Int): DataSet = {
    val fragment = ndarr.get(NDArrayIndex.interval(0, batchSize), NDArrayIndex.all(), NDArrayIndex.all())
    val fragmentMask = Nd4j.create(ndarr.shape()(0), 6)
    new DataSet(fragment, fragment, fragmentMask, fragmentMask)
  }

  override def cursor(): Int = throw new UnsupportedOperationException()
  override def totalExamples(): Int = ndarr.size(0)
  override def inputColumns(): Int = 1
  override def getPreProcessor: DataSetPreProcessor = throw new UnsupportedOperationException()
  override def setPreProcessor(preProcessor: DataSetPreProcessor): Unit = throw new UnsupportedOperationException()
  override def getLabels: util.List[String] = throw new UnsupportedOperationException()
  override def totalOutcomes(): Int = 1
  override def resetSupported(): Boolean = true
  override def reset() {}
  override def asyncSupported(): Boolean = false
  override def numExamples(): Int = ndarr.size(0)
  override def hasNext: Boolean = true
}

class OutputListener(test: INDArray) extends IterationListener {
  private var _invoked = false
  private var _iter = 1

  def invoke(): Unit = _invoked = true

  def invoked(): Boolean = _invoked

  def iterationDone(model: Model, iteration: Int): Unit = {
    model match {
      case _model: ComputationGraph =>
        val labels = _model.output(false, test)
        println("--------------- ITERATION DONE -------------")
        println(labels)
      case _ => //nop
    }
  }
}

val ds = new BatchedIterator(4)

val config = new NeuralNetConfiguration.Builder()
  .iterations(1)
  .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
  .learningRate(0.1)
  .weightInit(WeightInit.XAVIER)
  .updater(Updater.RMSPROP)
  .graphBuilder()
  .pretrain(false).backprop(true)
  .addInputs("input")
  .addLayer("encoderlstm", new GravesBidirectionalLSTM.Builder().nIn(1).nOut(2).activation(Activation.TANH).build(), "input")
  .addVertex("vertex", new LastTimeStepVertex("input"), "encoderlstm")
  .addLayer("dense", new DenseLayer.Builder().nIn(2).nOut(2).activation(Activation.IDENTITY).build(), "vertex")
  .addVertex("repeater", new DuplicateToTimeSeriesVertex("input"), "dense")
  .addLayer("decoderlstm", new GravesBidirectionalLSTM.Builder().nIn(2).nOut(2).activation(Activation.TANH).build(), "repeater")
  .addLayer("output", new RnnOutputLayer.Builder(LossFunction.MSE).nIn(2).nOut(1).activation(Activation.IDENTITY).build(), s"decoderlstm")
  .setOutputs("output")
  .build()

val testData = ds.next(2).getFeatures.dup()

val model = new ComputationGraph(config)
model.init()
model.setListeners(new OutputListener(testData)) // <- causes exception

model.fit(ds.next())

// model.output(false, testData) // <- works fine
```

**Comment by fahman** [comment]
Additional note: this same config works in 0.7.1 but throws in 0.8.0
